lib/helpers/model/discriminator.js

```diff
@@ -27,30 +27,32 @@ const CUSTOMIZABLE_DISCRIMINATOR_OPTIONS = {
  * @param {Schema} childSchema
  */
 function validateDiscriminatorSchemasForEncryption(parentSchema, childSchema) {
-  /**
-   * @param schema { Schema }
-   *
-   * Given a schema, yields **all** paths to values inside that schema, recursively iterating over any nested schemas.  This is
-   * intended for use with encryption, so nested arrays are not considered because encryption on values inside arrays is not supported.
-   *
-   * @returns { Iterable<string> }
-   */
-  function* allPaths(schema, prefix) {
-    for (const path of Object.keys(schema.paths)) {
-      const fullPath = prefix != null ? `${prefix}.${path}` : path;
-      if (schema.path(path).instance === 'Embedded') {
-        yield* allPaths(schema.path(path).schema, fullPath);
-      } else {
-        yield fullPath;
-      }
+  if (parentSchema.encryptionType() == null && childSchema.encryptionType() == null) return;
+
+  const allSharedNestedPaths = setIntersection(
+    allNestedPaths(parentSchema),
+    allNestedPaths(childSchema)
+  );
+
+  for (const path of allSharedNestedPaths) {
+    if (parentSchema._hasEncryptedField(path) && childSchema._hasEncryptedField(path)) {
+      throw new Error(`encrypted fields cannot be declared on both the base schema and the child schema in a discriminator. path=${path}`);
+    }
+
+    if (parentSchema._hasEncryptedField(path) || childSchema._hasEncryptedField(path)) {
+      throw new Error(`encrypted fields cannot have the same path as a non-encrypted field for discriminators. path=${path}`);
     }
   }
 
+  function* allNestedPaths(schema) {
```

Comment:
This is a neat use case for generators, but I think doing [...Object.keys(schema.paths), ...Object.keys(schema.singleNestedPaths)] would be more concise and more consistent with the rest of the codebase